File: AtCoder/ACP/BootCamp4b/Hard100_2/python/C_AB_Substrings.py
N = int(input())
S = [input() for _ in range(N)]

res = 0
cnt_BA = 0
cnt_B = 0
cnt_A = 0
for s in S:
    if s[0] == 'B' and s[-1] == 'A':
        cnt_BA += 1
    elif s[0] == 'B':
        cnt_B += 1
    elif s[-1] == 'A':
        cnt_A += 1
    for i in range(len(s) - 1):
        if s[i] == 'A' and s[i + 1] == 'B':
            res += 1
if cnt_BA > 0:
    res += cnt_BA - 1
    if cnt_B > 0:
        res += 1
        cnt_B -= 1
    if cnt_A > 0:
        res += 1
        cnt_A -= 1
res += min(cnt_B, cnt_A)
print(res)

File: AtCoder/ACP/BootCamp4b/Hard100_2/cpp/C_AB_Substrings.cpp
#include <bits/stdc++.h>
using namespace std;

int N;
vector<string> S;

void solve() {
  int cnt_BA = 0;
  int cnt_B = 0;
  int cnt_A = 0;
  int res = 0;
  for (int i = 0; i < N; ++i) {
    if (S[i][0] == 'B' && S[i][S[i].size() - 1] == 'A') {
      ++cnt_BA;
    }  else if (S[i][0] == 'B') {
      ++cnt_B;
    } else if (S[i][S[i].size() - 1] == 'A') {
      ++cnt_A;
    }
    for (int j = 0; j < S[i].size() - 1; ++j) {
      if (S[i][j] == 'A' && S[i][j + 1] == 'B') {
        ++res;
      }
    }
  }
  if (cnt_BA > 0) {
    res += cnt_BA - 1;
    if (cnt_A > 0) {
      ++res;
      --cnt_A;
    }
    if (cnt_B > 0) {
      ++res;
      --cnt_B;
    }
  }
  res += min(cnt_A, cnt_B);
  printf("%d\n", res);
}

int main() {
  cin >> N;
  S = vector<string>(N);
  for (int i = 0 ; i < N; ++i) {
    cin >> S[i];
  }
  solve();
}